### PR TITLE
Adding in filter

### PR DESCRIPTION
**Pull request by b-slim**
This PR is to fix https://github.com/druid-io/druid/issues/1887

### CODE REVIEW

**Comment by rasahner** [comment]
Should we add anything to the documentation about queries?

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  {
+    return values;
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+
+    byte[] dimensionBytes = StringUtils.toUtf8(dimension);
+    byte[] valueBytes = (values == null) ? new byte[]{} : StringUtils.toUtf8(values.toString());
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: it'd be more robust to explicitly build the cache key, rather than embedding the toString of a List.

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    return ByteBuffer.allocate(1 + dimensionBytes.length + valueBytes.length)
+                     .put(DimFilterCacheHelper.SELECTOR_CACHE_ID)
+                     .put(dimensionBytes)
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: there should be some kind of separator here, perhaps 0xff

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  private final String dimension;
+
+  @JsonCreator
+  public InDimFilter(@JsonProperty("dimension") String dimension, @JsonProperty("values") List<String> values)
+  {
+    this.values = values;
+    this.dimension = dimension;
+  }
+
+  public String getDimension()
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: Does this need `@JsonProperty`?

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  {
+    return dimension;
+  }
+
+  public List<String> getValues()
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: `@JsonProperty`

**Comment by gianm** [comment]
Yes, there should be docs or else nobody will know about the cool new feature! A test would be good too.

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java (hunk 2 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: Can you elaborate more why we need a separator ? i don't see it in the other filters...

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java (hunk 1 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: Can you explain what are the Pros and Cons

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: Usually its corner cases like people putting quotes and/or commas in their values.

The "safe" thing to do (and what we do elsewhere) is to add each value individually with an invalid UTF8 character (like `0xFF`) after each string.

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: Plus, what `toString()` prints out can vary depending on what `List` impl is used here.

Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: We want to guarantee that the cache key bytes and the thing we want to be the cache key have a one-to-one correspondence, to prevent false hits and false misses. Like @drcrallen said, the toString of a List does not satisfy that.

For example `Lists.newArrayList("a", "b", "c").toString()` and `Lists.newArrayList("a", "b, c").toString()` are both `[a, b, c]`.

Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: @gianm  and @drcrallen how about this impl

**Comment by b-slim** [comment]
Thanks for comments and let me know to rebase before merge.

File: docs/content/querying/filters.md [hunk 5]
```diff
@@ -133,6 +133,24 @@ Search filters can be used to filter on partial string matches.
...
 
+### In filter
+
+In filter can be used to express the following SQL query:
+
+```sql
+ SELECT COUNT(*) AS 'Count' FROM `table` WHERE `outlaw` IN ('Good', 'Bad', 'Ugly')
+```
+
+The grammar for a NOT filter is as follows:
```
Comment (rasahner) [inline docs/content/querying/filters.md]: grammar for an "in" filter  (yes?)

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    int index = 0;
+    for (String value : values) {
+      valuesBytes[index] = StringUtils.toUtf8(value);
+      valuesBytesSize += valuesBytes[index].length + 1;
+      ++index;
+    }
+
+    ByteBuffer filterCacheKey = ByteBuffer.allocate(1 + dimensionBytes.length + valuesBytesSize)
+                                          .put(DimFilterCacheHelper.IN_CACHE_ID)
+                                          .put(dimensionBytes);
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: @b-slim One more 0xff after the dimensionBytes and this looks good to me (or else the dimension name could get confused with the first value).

File: docs/content/querying/filters.md (hunk 5 above)
Comment (b-slim) [inline docs/content/querying/filters.md]: yes

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java (hunk 6 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: @gianm done

File: processing/src/main/java/io/druid/query/filter/InDimFilter.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+public class InDimFilter implements DimFilter
+{
+  private final List<String> values;
+  private final String dimension;
+
+  @JsonCreator
+  public InDimFilter(@JsonProperty("dimension") String dimension, @JsonProperty("values") List<String> values)
+  {
+    this.values = (values == null) ? Collections.<String>emptyList() : values;
+    this.dimension = dimension;
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/filter/InDimFilter.java]: `Preconditions.checkNotNull(dimension, "dimension")`

**Comment by gianm** [comment]
failed on `TestKafkaExtractionCluster.testSimpleRename:349 expected:<bar> but was:<null>`, bouncing

File: processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java [hunk 8]
```diff
@@ -1295,14 +1355,14 @@ public void testTimeseriesWithFilterOnNonExistentDimension()
...
-        .context(ImmutableMap.<String, Object>of("skipEmptyBuckets", "true"))
-        .build();
+                                  .dataSource(QueryRunnerTestHelper.dataSource)
+                                  .granularity(QueryRunnerTestHelper.dayGran)
+                                  .filters("bobby", "billy")
+                                  .intervals(QueryRunnerTestHelper.firstToThird)
+                                  .aggregators(QueryRunnerTestHelper.commonAggregators)
+                                  .postAggregators(Arrays.<PostAggregator>asList(QueryRunnerTestHelper.addRowsIndexConstant))
+                                  .context(ImmutableMap.<String, Object>of("skipEmptyBuckets", "true"))
+                                  .build();
```
Comment (indrekj) [inline processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java]: unrelated changes?

Comment (b-slim) [inline processing/src/test/java/io/druid/query/timeseries/TimeseriesQueryRunnerTest.java]: formatting

File: processing/src/main/java/io/druid/segment/filter/Filters.java [hunk 9]
```diff
@@ -93,6 +94,21 @@ public static Filter convertDimensionFilters(DimFilter dimFilter)
       final SpatialDimFilter spatialDimFilter = (SpatialDimFilter) dimFilter;
 
       filter = new SpatialFilter(spatialDimFilter.getDimension(), spatialDimFilter.getBound());
+    } else if (dimFilter instanceof InDimFilter) {
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/filter/Filters.java]: At this point, is it even possible to add a filter as an extension?

**Comment by drcrallen** [comment]
Overall looks good. Missing hash key and json serde tests though.

**Comment by b-slim** [comment]
@drcrallen added UTs ser desr and hash key

File: processing/src/main/java/io/druid/segment/filter/Filters.java (hunk 9 above)
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/filter/Filters.java]: i don't think it is straightforward..

**Comment by drcrallen** [comment]
I totally meant `getCacheKey`, sorry about that

File: processing/src/main/java/io/druid/segment/filter/Filters.java (hunk 9 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/filter/Filters.java]: Yeah, that would be a nice thing to have eventually

File: processing/src/test/java/io/druid/query/filter/InDimFilterSerDesrTest.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Test
+  public void testSerialization() throws IOException
+  {
+    final InDimFilter dimInFilter = new InDimFilter("dimTest", Arrays.asList("good", "bad"));
+    final String expectedInFilter = mapper.writeValueAsString(dimInFilter);
+    Assert.assertEquals(expectedInFilter, actualInFilter);
+  }
+}
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/filter/InDimFilterSerDesrTest.java]: newline

**Comment by gianm** [comment]
CI container died abrubtly, bouncing

**Comment by gianm** [comment]
looks good, :+1: after travis

File: processing/src/test/java/io/druid/query/filter/InDimFilterSerDesrTest.java (hunk 10 above)
Comment (b-slim) [inline processing/src/test/java/io/druid/query/filter/InDimFilterSerDesrTest.java]: done

**Comment by drcrallen** [comment]
I'd still like to see getCacheKey at least called in a UT, ideally tested for corner cases.

**Comment by b-slim** [comment]
@drcrallen i am not sure why getCacheKey should be tested, i have looked all the other filters i could not find any UT for getCacheKey, can you explain more why this is blocking ?

**Comment by gianm** [comment]
@b-slim I don't think it needs to block the PR but it would definitely be nice if you have time. I am :+1: with or without it. But test coverage is cool.

**Comment by himanshug** [comment]
I agree with @gianm , it is nice to have.

**Comment by drcrallen** [comment]
I asked both of these guys to comment as a sanity check and concede to their judgement.

**Comment by b-slim** [comment]
@drcrallen I have added the UT for that function. thanks everybody for the comments

**Comment by drcrallen** [comment]
Cool thanks!

**Comment by vogievetsky** [comment]
Hey @b-slim this is awesome. Thanks so much.

**Comment by drcrallen** [comment]
@b-slim : https://travis-ci.org/druid-io/druid/jobs/97728606#L4589 something weird in tests

**Comment by himanshug** [comment]
@drcrallen  that build is in passed state. whatz weird?
